File: leetcode/scala/213.house-robber-ii.scala
object Solution {
  def rob(nums: Array[Int]): Int = {
    // corner case
    if(nums.length == 0) return 0
    if(nums.length == 1) return nums(0)
      
    // pay attention to the slice edge.
    val first = simple_rob(nums.slice(0, nums.length - 1))
    val second = simple_rob(nums.slice(1, nums.length))
    if(first > second) return first
    second
  }

  def simple_rob(nums: Array[Int]): Int = {
    if (nums.length == 0) {
      0
    }
    if (nums.length == 1) {
      return nums(0)
    }
    if (nums.length == 2) return {
      if (nums(0) > nums(1)) nums(0)
      else nums(1)
    }
    val res = new Array[Int](nums.length)
    res(0) = nums(0)
    res(1) = if (nums(0) > nums(1)) nums(0) else nums(1)

    for (i <- Range(2, nums.length)) {
      res(i) = if (res(i - 1) > (res(i - 2) + nums(i))) {
        res(i - 1)
      }
      else {
        (res(i - 2) + nums(i))
      }
    }

    res(res.length - 1)
  }
}

File: leetcode/js/213.house-robber-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
    if (nums.length == 0) return 0;
    if (nums.length == 1) return nums[0];
    if (nums.length == 2) return Math.max(nums[0], nums[1]);

    return Math.max(_rob(nums.slice(0, nums.length - 1)), _rob(nums.slice(1)))
};

/**
 * @param {number[]} nums
 * @return {number}
 */
var _rob = function (nums) {
    if (nums.length == 0) return 0;
    if (nums.length == 1) return nums[0];
    if (nums.length == 2) return Math.max(nums[0], nums[1]);

    let dp = [nums[0], Math.max(nums[0], nums[1])];

    for (let i = 2; i < nums.length; i++) {
        const element = nums[i];
        dp.push(Math.max(element + dp[i - 2], dp[i - 1]));
    }

    return dp[dp.length - 1];

};